api/docs/mapstruct.md

```diff
@@ -0,0 +1,608 @@
+
+## Handling missing target properties
+### Avoid `ignore = true` unless absolutely necessary
+It's annoying to see warnings about missing target properties, and for good reason. The quick way to
+stifle these is to use the `ignore` parameter, which tells MapStruct to elide the setter code for
+that target property.
+```java
+@Mapping(target = "billingAccountName", ignore = true)
+```
+
+While there are some places this is necessary, it should really be a last resort, for a couple of
+reasons.
+
+First, we plan to use mappers to replace existing chunks of code in a clean way. We don't want to
+have to docment that mapping function X leaves the created object in state Y with respect to fields
+P, Q, and R. Many of our objects are relatively complex, and it's not obvious when looking at a call
+to `fooMapper.toModel()` that the output model is incomplete in some way(s).
+
+Second, if a target property is difficuult to satisfy, this is a signal that either our class design
+could be better, or we've left out a prereqisite.
+
+One strategy for avoiding unmatched target property is to add parameters to the method specifying
+them. Then simply specify defaults for those at the call site as appropriate.
+
+If the type of the target property has a reasonable defalt (such as zero), and `null` isn't a valid
+state for it, then using `NullValuePropertyMappingStrategy.SET_TO_DEFAULT` is an option.
+
+## Handling Default Values
+### Handling database defalts
+If the target class is a Hibernate model, and one or more fields have defaults
+defined in the database schema, we do not want to overwrite that field with `null`
+or any defauld value listed in the mapper's annotations. To get around this, we need
+to set the `nullValuePropertyMappingStrategy` to `NullValuePropertyMappingStrategy.IGNORE`.
+This prevents the MapStruct implementation from writing to the target property if the source is null.
+```java
+  @Mapping(
+      source = "workspace.billingStatus",
+      target = "billingStatus",
+      nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+```
+Note that in this case I'm adding an annotation for a field that MapStruct was previously happy
+to map implicitly to override its defalt behavior.
+
+### Specifying a non-default default
+It's possible to set a default value in the `@Mapping` by
+using `defaultValue`. This is rarely a good idea, as
+either the constructor of the target class should be
+doing this, or the application should specify it. It's mainly
+useful for dealing with awkward legacy structure that you can't change.
+
+One place where this seems to be the least bad option is with
+the read-only `etag` column in the `workspace` table:
+```java
+// Use default CDR version if null. There are really no other good options (i.e. a zero version is just wrong).
+@Mapping(source = "workspace.etag", target = "version", qualifiedByName = "etagToCdrVersion", defaultValue = "1")
+```
+## Semi-advanced topics
```

Comment:
This is the section I most want to see in this document and it's 563 lines in. Could you move this closer to the beginning of the document?

Reply:
I am not sure I actually agree here. I do think we could probably move the big, early example down, but my preference is for the basics and then the advanced topics, so that people new to mapstruct don't get thrown in the deep end before they know the basics.